docs/tutorials/forum-application/3.send-message.md

```diff
@@ -0,0 +1,381 @@
+
+After that, it iterates through the remaining transactions (starting from the index after the last ban transaction) and
+parses each transaction using the `model.ParseMessage` function. If any banned user attempts to send a transaction,
+it rejects the proposal.
+
+Finally, if there are no banned users found in the transactions, it accepts the proposal.
+
+*Tip: The function `ProcessProposal` is used by state replication to indicate to the application to process the tx.
+The application can process a tx in accordance to the logic defined by the application. Although the application can
+perform 'optimisitic execeution', the application is not mandated to do so.*
+
+Note: You will learn about different packages and functions like `isBanTx` in the upcoming sections. In this section
+you will learn about `app.go` file only.
+
+## FinalizeBlock
+
+The `FinalizeBlock` method is used to finalize a block in the blockchain. It is called by CometBFT after all the transactions
+in a block have been processed and the block is ready to be added to the blockchain.
+
+The `FinalizeBlock` method takes in a `FinalizeBlockRequest` object, which contains information about the block being finalized.
+It performs any necessary processing or validation on the block, such as updating the application state, or performing
+additional computations.
+
+After processing the block, the method returns a `FinalizeBlockResponse` object, which typically includes the results
+of the block finalization process, such as the transaction results, validator set updates, consensus parameters updates
+or the new hash of the application state.
+
+**Following is the code for the `FinalizeBlock` function:**
+
+```go
+// Deliver the decided block to the Application
+func (app *ForumApp) FinalizeBlock(_ context.Context, req *abci.FinalizeBlockRequest) (*abci.FinalizeBlockResponse, error) {
+	fmt.Println("Executing Application FinalizeBlock")
+
+	// Iterate over Tx in current block
+	app.onGoingBlock = app.state.DB.GetDB().NewTransaction(true)
+	respTxs := make([]*abci.ExecTxResult, len(req.Txs))
+	finishedBanTxIdx := len(req.Txs)
+	for i, tx := range req.Txs {
+		var err error
+
+		if isBanTx(tx) {
+			banTx := new(model.BanTx)
+			err = json.Unmarshal(tx, &banTx)
+			if err != nil {
+				respTxs[i] = &abci.ExecTxResult{Code: CodeTypeEncodingError}
+			} else {
+				err := UpdateOrSetUser(app.state.DB, banTx.UserName, true, app.onGoingBlock)
+				if err != nil {
+					panic(err)
+				}
+				respTxs[i] = &abci.ExecTxResult{Code: CodeTypeOK}
+			}
+		} else {
+			finishedBanTxIdx = i
+			break
+		}
+	}
+
+	for idx, tx := range req.Txs[finishedBanTxIdx:] {
+		// From this point on, there should be no BanTxs anymore
+		// If there is one, ParseMessage will return an error as the
+		// format of the two transactions is different.
+		msg, err := model.ParseMessage(tx)
+		i := idx + finishedBanTxIdx
+		if err != nil {
+			respTxs[i] = &abci.ExecTxResult{Code: CodeTypeEncodingError}
```

Comment:
Again, this should never happen. If it happens, an important bug exists in the application, so we should probably panic here, not continue.